devdoc/slist.md

```diff
@@ -0,0 +1,171 @@
+# slist requirements
+================
+
+## Overview
+
+`slist` is module that provides the functionality of a singly linked list, allowing its user to add, remove and iterate the list elements.
+
+## Exposed API
+
+```c
+typedef struct SLIST_ENTRY_TAG
+{
+    struct SLIST_ENTRY_TAG *next;
+} SLIST_ENTRY, *PSLIST_ENTRY;
+
+typedef bool (*SLIST_MATCH_FUNCTION)(PSLIST_ENTRY list_entry, const void* match_context);
+typedef bool (*SLIST_CONDITION_FUNCTION)(PSLIST_ENTRY list_entry, const void* match_context, bool* continue_processing);
+typedef int (*SLIST_ACTION_FUNCTION)(PSLIST_ENTRY list_entry, const void* action_context, bool* continue_processing);
+
+MOCKABLE_FUNCTION(, bool, slist_initialize, PSLIST_ENTRY, list_head);
+MOCKABLE_FUNCTION(, bool, slist_is_empty, const PSLIST_ENTRY, list_head);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, slist_add, PSLIST_ENTRY, list_head, PSLIST_ENTRY, list_entry);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, slist_add_head, PSLIST_ENTRY, list_head, PSLIST_ENTRY, list_entry);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, slist_remove, PSLIST_ENTRY, list_head, PSLIST_ENTRY, list_entry);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, slist_remove_head, PSLIST_ENTRY, list_head);
+MOCKABLE_FUNCTION(, PSLIST_ENTRY, slist_find, PSLIST_ENTRY, list_head, SLIST_MATCH_FUNCTION, match_function, void*, match_context);
+MOCKABLE_FUNCTION(, int, slist_remove_if, PSLIST_ENTRY, list_head, SLIST_CONDITION_FUNCTION, condition_function, void*, match_context);
+MOCKABLE_FUNCTION(, int, slist_for_each, PSLIST_ENTRY, list_head, SLIST_ACTION_FUNCTION, action_function, void*, action_context);
+
+```
+## Usage
+
+`SLIST_ENTRY` does not store any data. Instead, `SLIST_ENTRY` is placed inside a struct:
+
+```c
+typedef struct DATA_TAG
+{
+    int num1;
+    int num2;
+    SLIST_ENTRY anchor;
+} DATA;
+```
+
+The parent struct of a list entry can be obtained using `CONTAINING_RECORD`:
+
+```
+PSLIST_ENTRY list_entry = slist_remove_head(list_head);
+DATA* data = CONTAINING_RECORD(list_entry, DATA, anchor);
+```
+
+### slist_initialize
+```c
+bool slist_initialize(PSLIST_ENTRY list_head);
+```
+
+If `list_head` is `NULL`, `slist_initialize` shall fail and return `NULL`. 
```

Comment:
Nitpick: shall fail and return `NULL`.  It should be false. 

Here and other places.